extensions/amp-story/1.0/amp-story-parallax-service.js

```diff
@@ -0,0 +1,430 @@
+/**
+ * Copyright 2019 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {Services} from '../../../src/services';
+import {dev} from '../../../src/log';
+import {escapeCssSelectorIdent} from '../../../src/css';
+import {findIndex} from '../../../src/utils/array';
+import {isExperimentOn} from '../../../src/experiments';
+import {listen} from '../../../src/event-helper';
+import {mapRange, sum} from '../../../src/utils/math';
+import {registerServiceBuilder} from '../../../src/service';
+import {resetStyles, setImportantStyles, setStyles} from '../../../src/style';
+import {toArray} from '../../../src/types';
+
+const SMOOTHING_PTS = 4;
+const PERSPECTIVE = 1000;
+const DEFAULT_LAYER_SPACING = 1;
+const DEFAULT_FARTHEST_SCALE = 1.3;
+const DEFAULT_NEAREST_SCALE = 0.8;
+const AMPLIFICATION_FACTOR = 40;
+
+const LAYER_SPACING_ATTR = 'parallax-fx-layer-spacing';
+const NEAREST_SCALE_ATTR = 'parallax-fx-nearest-scale';
+const FARTHEST_SCALE_ATTR = 'parallax-fx-farthest-scale';
+const PARALLAX_FX_ATTR = 'parallax-fx';
+const NO_PARALLAX_FX_ATTR = 'no-parallax-fx';
+const ORIGIN_LAYER_ATTR = 'parallax-fx-origin-layer';
+
+/**
+ * Util function to retrieve the parallax service. Ensures we can retrieve the
+ * service synchronously from the amp-story codebase without running into race
+ * conditions.
+ * @param  {!Window} win
+ * @param {!Element=} opt_story
+ * @return {?AmpStoryParallaxService}
+ */
+export const getParallaxService = (win, opt_story) => {
+  let service = Services.storyParallaxService(win);
+
+  if (!service && opt_story) {
+    service = new AmpStoryParallaxService(win, dev().assertElement(opt_story));
+    registerServiceBuilder(win, 'story-parallax', () => service);
```

Comment:
Why not just register the service builder to begin with? It's only going to be built once.

Reply:
From offline discussion: this should be divided into two methods but since this is a pattern that occurs throughout the services in `amp-story` let's not make an exception here